### PR TITLE
refactor(npm): extract replacement to a private function

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes:

Extract dependency replacement into a separate function instead of repeating the logic

## Context:

https://github.com/renovatebot/renovate/pull/5558#discussion_r678058083

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/manager/npm/update/dependency/index.ts [hunk 1]
```diff
@@ -3,6 +3,55 @@ import { logger } from '../../../../logger';
 import { matchAt, replaceAt } from '../../../../util/string';
 import type { UpdateDependencyConfig } from '../../../types';
 
+function replaceAsString(
+  parsedContents: any,
```
Comment (viceice) [inline lib/manager/npm/update/dependency/index.ts:7]: I think we already have a type for this, so we won't need `any` here.

Comment (JamieMagee) [reply lib/manager/npm/update/dependency/index.ts]: `PackageJson`?

File: lib/manager/npm/update/dependency/index.ts [hunk 2]
```diff
@@ -1,8 +1,58 @@
...
+  depName: string,
+  oldVersion: string,
+  newValue: string
+): string | null {
+  // Update the file = this is what we want
+  // eslint-disable-next-line no-param-reassign
+  if (depName === oldVersion) {
+    delete Object.assign(parsedContents[depType], {
+      [newValue]: parsedContents[depType][oldVersion],
+    })[oldVersion];
```
Comment (viceice) [inline lib/manager/npm/update/dependency/index.ts:20]: This looks strange to me, also coverage is missing.

Comment (JamieMagee) [reply lib/manager/npm/update/dependency/index.ts:20]: 🤦 You're right. This is from my other work in #5558. I directly copy-pasted this function from there and didn't check more closely.

File: lib/manager/npm/update/dependency/index.ts [hunk 3]
```diff
@@ -1,8 +1,53 @@
...
+function replaceAsString(
+  parsedContents: PackageJson,
+  fileContent: string,
+  depType: string,
+  depName: string,
+  oldVersion: string,
+  newValue: string
+): string | null {
+  // Update the file = this is what we want
+  if (depType === 'resolutions') {
```
Comment (viceice) [inline lib/manager/npm/update/dependency/index.ts:16]: Why this condition, looks like both branches are doing the same?

File: lib/manager/npm/update/dependency/index.ts [hunk 4]
```diff
@@ -1,8 +1,53 @@
...
+        newString
+      );
+      // Compare the parsed JSON structure of old and new
+      if (dequal(parsedContents, JSON.parse(testContent))) {
+        return testContent;
+      }
+    }
+  }
+  // istanbul ignore next: not possible to get here
+  return null;
```
Comment (viceice) [inline lib/manager/npm/update/dependency/index.ts:48]: Should we throw an error Here? 🤔

Comment (JamieMagee) [reply lib/manager/npm/update/dependency/index.ts:48]: Good question. If I am reading the old code path right, if we got here, we would silently ignore it. Right?

**Comment by viceice** [comment]
needs code coverage

File: lib/manager/npm/update/dependency/index.ts [hunk 5]
```diff
@@ -1,8 +1,47 @@
...
+        searchString,
+        newString
+      );
+      // Compare the parsed JSON structure of old and new
+      if (dequal(parsedContents, JSON.parse(testContent))) {
+        return testContent;
+      }
+    }
+  }
+  throw new Error();
```
Comment (rarkins) [inline lib/manager/npm/update/dependency/index.ts:42]:
```suggestion
  // istanbul ignore next
  throw new Error();
```

**Review by rarkins — APPROVED** [review]

**Pull request merged**